File: src/orchestrator-roadmap.ts
import { analyzeRoadmap, type RoadmapStatus } from "./spec-roadmap.js";
import type { CuriousState } from "./types.js";

/** At untilDone start: skip the phase loop when every roadmap task is already [x]. */
export function shouldSkipUntilDoneLoop(roadmapStatus: RoadmapStatus): boolean {
  return roadmapStatus.complete;
}

/** After a phase: stop untilDone when sync finished and spec roadmap is complete. */
export function shouldStopUntilDoneAfterPhase(
  state: CuriousState,
  specBody: string,
): boolean {
  const last = state.history[state.history.length - 1];
  if (last?.phase !== "sync" || last.status !== "finished") {
    return false;
  }
  return analyzeRoadmap(specBody).complete;
}

File: python/src/curious/orchestrator_roadmap.py
from __future__ import annotations

from curious.spec_roadmap import RoadmapStatus, analyze_roadmap
from curious.types import CuriousState, CycleRecord


def should_skip_until_done_loop(roadmap_status: RoadmapStatus) -> bool:
    return roadmap_status.complete


def should_stop_until_done_after_phase(
    state: CuriousState, spec_body: str
) -> bool:
    if not state.history:
        return False
    last = state.history[-1]
    if last.phase != "sync" or last.status != "finished":
        return False
    return analyze_roadmap(spec_body).complete
